### PR TITLE
proposal: Introduce event driver waiting tooling for e2e tests

### PR DESCRIPTION
**Pull request by EdDev**
The existing e2e tests at kubevirt/kubevirt is frequently using
polling to wait for a specific object state (e.g. VMI running).

This proposal suggests to use an event driven approach instead.

---

I got some requests for ref usage if this `watch` tooling.
[Here](https://github.com/kubernetes/kubectl/search?q=watchtools.Until) are the ones from kubectl.
(thank you @qinqon for finding it)

### CODE REVIEW

File: design-proposals/e2e-waitfor-tooling.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+to setup and manage them.
+Its users are left with just adding basic logic regarding the
+conditions and possibly filtering (e.g. label selectors).
+
+The main function in use is `Until`, supporting a timeout, aborting
+and conditions.
+
+Example (waiting for namespace deletion):
+```
+_, err := k8swatchtools.Until(ctx, namespace.ResourceVersion, w, func(event k8swatch.Event) (bool, error) {
```
Comment (maiqueb) [inline design-proposals/e2e-waitfor-tooling.md:133]: this signature is wrong isn't it ?

Comment (EdDev) [reply design-proposals/e2e-waitfor-tooling.md:133]: Can you help me detect the problem?
I took it from here: https://github.com/kiagnose/kiagnose/blob/3bb666abd01ddaf40941e7bd9f06443ef945e60f/kiagnose/internal/checkup/namespace/namespace.go#L68

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thanks for your proposal. I bet it would make the code cleaner, more robust and ever so slightly faster.

File: design-proposals/e2e-waitfor-tooling.md (hunk 1 above)
Comment (maiqueb) [reply design-proposals/e2e-waitfor-tooling.md:133]: well, that's a call. I was expecting the signature (probably my bad, this is an example), which conveys more info. I have no idea what that `w` in there is - if I had the signature, I would know.

for convenience: https://github.com/kubernetes/client-go/blob/53da6da2bcb15dc987fa5b19eb33c4793e574c99/tools/watch/until.go#L106

File: design-proposals/e2e-waitfor-tooling.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+```
+
+## Phases
+
+- PoC: Introduce `Until` into the e2e tests at kubevirt/kubevirt,
+  replacing a single usage of one of the polling implementations.
+- Start migrating a group of tests to this new method and monitor
+  it for a period of several weeks (for stability).
+- Once (and if) the stability monitoring finishes with satisfying
+  results, start migrating other tests to it. This should include
```
Comment (davidvossel) [inline design-proposals/e2e-waitfor-tooling.md:152]: For existing stable tests, I see little value in transitioning them to use the event pattern unless something else is being done in that area already.

So, when it comes to old tests, I'd suggest transitioning to the event pattern only if existing flakiness is being addressed. Otherwise if a test is stable already, we're getting diminished returns on this effort... and run the risk of actually introducing new instability.

Basically, I'd encourage us to prioritize old tests that are flaky and not take random tests and start transitioning them to the event pattern.

File: design-proposals/e2e-waitfor-tooling.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+The e2e tests can be classified as applications that integrate with the
+Kubernetes cluster, and therefore, could (and should) use the same
+patterns expected from production applications.
+
+## Goals
+
+- Use event driven approach to wait for a condition. Encourage it over
+  the polling approach.
+- Increase e2e test stabilization.
```
Comment (davidvossel) [inline design-proposals/e2e-waitfor-tooling.md:50]: to be clear, this question is not an argument against using an event watching pattern

I understand that watching is more efficient that pooling, but for our test suite, do we have evidence that our existing poll logic is causing instability?

Comment (xpivarc) [reply design-proposals/e2e-waitfor-tooling.md:50]: @dankenigsberg I am not sure the example is relevant for this discussion. (Side note: The test could use eventually as we expect that eventually, we will fail to create the VM after deleting the CDI)

Comment (davidvossel) [reply design-proposals/e2e-waitfor-tooling.md:50]: yeah, in this specific case I don't think we have any objects to watch in order to ensure the kubevirt components have transitioned to not using CDI anymore. Reporting that on an object (like the kv object as a condition) might be a good improvement in the future. 

That isn't exactly supporting evidence for instability related to polling vs event watching within the context of using informers vs GET requests though.

Comment (dankenigsberg) [reply design-proposals/e2e-waitfor-tooling.md:50]: Yeah, my URL is relevant only in a very convoluted way (a feature should be written in such a way that it can be tested with a watch). A better example (still theoretical, no proof for the assertion in the doc) is every time we use `Consistently` in a test. Without a watch we cannot really tell that the state of the production resource was indeed consistent.
